### PR TITLE
added jsonapi task and templates

### PR DESCRIPTION
**Pull request by alexjp**
Copied phoenix.gen.json and changed it to use jsonapi.

Changed controller and views accordingly.

### CODE REVIEW

**Comment by jeregrine** [comment]
@alexjp my only concern on the face is making sure we stay up to date with all phoenix changes. I will talk to chris real quick

File: priv/templates/jsonapi/controller_test.exs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  test "does not update chosen resource and renders errors when data is invalid", %{conn: conn} do
+    <%= singular %> = Repo.insert! %<%= alias %>{}
+    conn = put conn, <%= singular %>_path(conn, :update, <%= singular %>), <%= singular %>: @invalid_attrs
+    assert json_response(conn, 422)["errors"] != %{}
+  end
+
+  test "deletes chosen resource", %{conn: conn} do
+    <%= singular %> = Repo.insert! %<%= alias %>{}
+    conn = delete conn, <%= singular %>_path(conn, :delete, <%= singular %>)
+    assert json_response(conn, 200)["data"]["id"]
```
Comment (jeregrine) [inline priv/templates/jsonapi/controller_test.exs]: this should be 202 http://jsonapi.org/format/#crud-deleting-responses

File: priv/templates/jsonapi/controller_test.exs [hunk 2]
```diff
@@ -0,0 +1,62 @@
+defmodule <%= module %>ControllerTest do
```
Comment (jeregrine) [inline priv/templates/jsonapi/controller_test.exs]: we should prolly make sure this follows the req response codes from http://jsonapi.org/format/

File: priv/templates/jsonapi/controller.ex [hunk 3]
```diff
@@ -0,0 +1,51 @@
+defmodule <%= module %>Controller do
```
Comment (jeregrine) [inline priv/templates/jsonapi/controller.ex]: ditto here re- response codes http://jsonapi.org/format/#crud-deleting-responses

**Comment by jeregrine** [comment]
Looks good. Couple comment's re-jsonapi response codes. Thanks a ton for the support!

**Comment by jeregrine** [comment]
Also @alexjp if we could add to the README or docs on how to use this (just an example maybe) it would be helpful!

Once the proper codes are sent by the controller I will merge this :+1: thank you so much. 

I will prolly also work on getting some of my helper functions into this as I add them

**Comment by alexjp** [comment]
@jeregrine sorry for the lack of correctness on the pull request, it was done while working day for a project.

I will test more and more, and update things.

Also I can try on a usage doc, since I am a elixir newbie, I can probably pose what should be necessary.

File: priv/templates/jsonapi/view.ex [hunk 4]
```diff
@@ -0,0 +1,19 @@
+defmodule <%= module %>View do
+  use <%= base %>.Web, :view
+
+  use JSONAPI.PhoenixView
+
+  def type, do: "<%= singular %>"
+
+  def attributes(model) do
+    Map.take(model, [:id, :created_at])
```
Comment (jeregrine) [inline priv/templates/jsonapi/view.ex]: Should probably also generate this similar to https://github.com/phoenixframework/phoenix/blob/master/priv/templates/model/model.ex

**Comment by amutz** [comment]
Just want to say :+1: to this PR.  I'd find a lot of value in this generator.

**Comment by alexjp** [comment]
@amutz in latest commit fix, they should "just work" ( was wrongly searching for them in phoenix , not jsonapi )

File: priv/templates/jsonapi/controller.ex [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  plug :scrub_params, <%= inspect singular %> when action in [:create, :update]
+
+  def index(conn, _params) do
+    <%= plural %> = Repo.all(<%= alias %>)
+    render conn, "index.json", %{data: <%= plural %>, params: _params}
+  end
+
+  def create(conn, %{<%= inspect singular %> => <%= singular %>_params}) do
+    changeset = <%= alias %>.changeset(%<%= alias %>{}, <%= singular %>_params)
```
Comment (jeregrine) [inline priv/templates/jsonapi/controller.ex]: this needs to be 

```
Map.get(signular_params, "data")
```

File: priv/templates/jsonapi/controller.ex [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+  def show(conn, %{"id" => id}) do
+    <%= singular %> = Repo.get!(<%= alias %>, id)
+    render conn, "show.json", %{data: <%= singular %>}
+  end
+
+  def update(conn, %{"id" => id, <%= inspect singular %> => <%= singular %>_params}) do
+    <%= singular %> = Repo.get!(<%= alias %>, id)
+    changeset = <%= alias %>.changeset(<%= singular %>, <%= singular %>_params)
```
Comment (jeregrine) [inline priv/templates/jsonapi/controller.ex]: ditto

File: priv/templates/jsonapi/view.ex (hunk 4 above)
Comment (alexjp) [inline priv/templates/jsonapi/view.ex]: talking about the : 
```
for {k, _} <- attrs do %>
```

how do i get that attrs ?

Comment (jeregrine) [inline priv/templates/jsonapi/view.ex]: You need to take a look at the phoenix generators here https://github.com/phoenixframework/phoenix/blob/master/lib/mix/tasks/phoenix.gen.model.ex#L56 

The default gen.contrller.json that this is based on calls the model generator too

**Comment by bbhoss** [comment]
This looks cool. What else is needed to get this in? Need help?

File: lib/mix/tasks/phoenix.gen.jsonapi.ex [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+      mix phoenix.gen.json User users name:string age:integer
+
+  The first argument is the module name followed by
+  its plural name (used for resources and schema).
+
+  The generated resource will contain:
+
+    * a model in web/models
+    * a view in web/views
```
Comment (jeregrine) [inline lib/mix/tasks/phoenix.gen.jsonapi.ex]: jsonapi view

File: lib/mix/tasks/phoenix.gen.jsonapi.ex [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    end
+  end
+
+  defp validate_args!(_) do
+    raise_with_help
+  end
+
+  defp raise_with_help do
+    Mix.raise """
+    mix phoenix.gen.json expects both singular and plural names
```
Comment (jeregrine) [inline lib/mix/tasks/phoenix.gen.jsonapi.ex]: jsonapi

File: lib/mix/tasks/phoenix.gen.jsonapi.ex [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    of the generated resource followed by any number of attributes:
+
+        mix phoenix.gen.json User users name:string
```
Comment (jeregrine) [inline lib/mix/tasks/phoenix.gen.jsonapi.ex]: jsonapi

File: mix.exs [hunk 10]
```diff
@@ -30,8 +30,8 @@ defmodule JSONAPI.Mixfile do
   # Type `mix help deps` for more examples and options
   defp deps do
     [
-      {:phoenix, "~> 0.13"},
-      {:ecto, "~> 0.11"},
+      {:phoenix, "~> 0.14"},
```
Comment (jeregrine) [inline mix.exs]: this is two versions behind now. We'll wanna update the whole generator to match the latest changes in that. 

This is gonna be the tricky part, keeping in sync.

File: priv/templates/jsonapi/controller.ex [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    <%= plural %> = Repo.all(<%= alias %>)
+    render conn, "index.json", %{data: <%= plural %>, params: _params}
+  end
+
+  def create(conn, %{<%= inspect singular %> => <%= singular %>_params}) do
+    changeset = <%= alias %>.changeset(%<%= alias %>{}, <%= singular %>_params)
+
+    if changeset.valid? do
+      <%= singular %> = Repo.insert!(changeset)
+      render conn, "show.json", %{data: <%= singular %>}
```
Comment (jeregrine) [inline priv/templates/jsonapi/controller.ex]: this response doesn't match JSONAPI spec http://jsonapi.org/format/#crud-creating

File: priv/templates/jsonapi/controller.ex [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    else
+      conn
+      |> put_status(:unprocessable_entity)
```
Comment (jeregrine) [inline priv/templates/jsonapi/controller.ex]: this does not match JSONAPI spec

File: priv/templates/jsonapi/controller.ex [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    render conn, "show.json", %{data: <%= singular %>}
+  end
+
+  def update(conn, %{"id" => id, <%= inspect singular %> => <%= singular %>_params}) do
+    <%= singular %> = Repo.get!(<%= alias %>, id)
+    changeset = <%= alias %>.changeset(<%= singular %>, <%= singular %>_params)
+
+    if changeset.valid? do
+      <%= singular %> = Repo.update!(changeset)
+      render conn, "show.json", %{data: <%= singular %>}
```
Comment (jeregrine) [inline priv/templates/jsonapi/controller.ex]: does not match jsonapi response format http://jsonapi.org/format/#crud-updating-responses

File: priv/templates/jsonapi/controller.ex [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+    else
+      conn
+      |> put_status(:unprocessable_entity)
```
Comment (jeregrine) [inline priv/templates/jsonapi/controller.ex]: does not match jsonapi response format http://jsonapi.org/format/#errors

File: priv/templates/jsonapi/controller_test.exs (hunk 2 above)
Comment (jeregrine) [inline priv/templates/jsonapi/controller_test.exs]: these tests will need to be updated to follow the format.

**Comment by jeregrine** [comment]
@bbhoss @alexjp looking better but we are very far from being JSONAPI compliant out of the box with this generator.

File: mix.exs (hunk 10 above)
Comment (bbhoss) [inline mix.exs]: Indeed. Hopefully with 1.0 things will stabilize a bit.

**Comment by bbhoss** [comment]
Thanks. I think I am going to start using this later this week. I'll see how I end up creating things and try to use that to inform development of the generators.

**Comment by alexjp** [comment]
@jeregrine  i will look at this today and try to solve it.

**Comment by alexjp** [comment]
@jeregrine updated so that the view uses attrs.

Who do i make so that in the last one, do not insert the "," ?

**Comment by alexjp** [comment]
@jeregrine updated phoenix and phoenix_ecto is mix.iex.

is it better ecto or phoenix_ecto ?

**Comment by jeregrine** [comment]
@alexjp going to close this since we'll need to restart this anyways. Sorry for not merging sooner.

**Pull request closed without merging**